crates/formats/src/param.rs

```diff
@@ -0,0 +1,561 @@
+                is_64_bit: header.is_64_bit(),
+                is_unicode: header.is_unicode(),
+            });
+        }
+
+        let mut detected_strings_offset: Option<u64> = None;
+        let mut detected_row_size: Option<u64> = None;
+
+        // Parse the param type string (expecting ASCII, so failing if not utf8 is fine)
+        let param_type = if header.is_long_param_type() {
+            // SAFETY: union access is always safe for FromBytes
+            let inlined = unsafe { &header.param_type_block.param_type_buf };
+            let inlined_nt = &inlined[..inlined
+                .iter()
+                .position(|b| *b == 0)
+                .unwrap_or(inlined.len())];
+
+            str::from_utf8(inlined_nt).map_err(|_| ParamParseError::InvalidData)?
+        } else {
+            // SAFETY: union access is always safe for FromBytes
+            let offset: u64 = unsafe { header.param_type_block.offset.param_type_offset }.into();
+            let strings = data
+                .get(offset as usize..)
+                .ok_or(ParamParseError::InvalidData)?;
+
+            // This is going to be the best guess for the strings offset
+            detected_strings_offset = Some(offset);
+
+            CStr::from_bytes_until_nul(strings)
+                .ok()
+                .and_then(|cstr| cstr.to_str().ok())
+                .ok_or(ParamParseError::InvalidData)?
+        };
+
+        // Parse the row descriptors
+        let row_descriptors = RowDescriptor::<T>::slice_from_prefix(
+            data.get(header.header_size()..)
+                .ok_or(ParamParseError::InvalidData)?,
+            header.row_count(),
+        )
+        .ok_or(ParamParseError::InvalidData)?
+        .0;
+
+        // If we have two rows, try to use them to guess the row size
+        if row_descriptors.len() >= 2 {
+            let diff = (&row_descriptors[1])
+                .data_offset
+                .into()
+                .checked_sub(row_descriptors[0].data_offset.into())
+                .ok_or(ParamParseError::InvalidData)?;
+            detected_row_size = Some(diff);
+        }
+        if row_descriptors.len() >= 1 {
+            // If the row name offset is set, this will be the string offset
+            // provided it is not already set (as we have one row).
+            // Otherwise, fall back to the (very unreliable) field in the header.
+            let row_0_name_ofs = row_descriptors[0].name_offset.into();
+            if row_0_name_ofs != 0 {
+                detected_strings_offset.get_or_insert(row_0_name_ofs);
+            } else if header.strings_offset != U32::ZERO {
+                detected_strings_offset.get_or_insert(header.strings_offset.into());
+            }
+            // If we have a single row and the strings offset is known,
+            // then the row size will be the difference between the row data and strings
+            if let Some(o) = detected_strings_offset {
```

Comment:
Nit: no single letter variables please, `offset` is fine.

Reply:
Done.